### ISSUE TITLE
Classifier information view should show subtype hierarchy

### ISSUE THREAD

**Issue by lwrage** [osate/osate2#2436]
## Summary

The classifier view should also show component types and implementations that extend the classifier that is currently in focus.

## Expected and Current Behavior

Currently the view only shows types and implementations that are extended by the classifier in focus, i.e., the supertype hierarchy. Subtypes should also be displayed.

Compare the Java type hierarchy view, which has type hierarchy, supertye hierarchy, and subtype hierarchy.

![image](https://user-images.githubusercontent.com/750311/93265075-12b0ce00-f776-11ea-9c91-ed9b020aa32a.png)

## Environment

* **OSATE Version**: 2.8.0
* **Operating System**: all

**Comment by AaronGreenhouse** [comment]
Constructing the subtype hiearchy rooted at the selected classifier is harder than getting the ancestors: Need to search the workspace for all references to the classifier and then see if the reference is in an `extends` or implementation clause.  Need to identify the following:

* type extension: i.e., `system A extends B` — type A extends type B

* type implementation: i.e., `system implementation A.i` — implementation A.i implements type A

* implementation extension: i.e, `system implementation A.i extends C.k` — implementation A.i extends C.k (and also implements type A)

Need to use the class `org.osate.search.AadlFinder`.  See `FindUnusedClassifiersAnalysis` for example of usage.

**Comment by AaronGreenhouse** [comment]
At first I thought that this was going to affect the changes made for Issue #2435 (updating when the file is changed).  But this is not true.  **I must always watch the ancestors of the selected classifier for changes** to update the RHS of the view.  Even when the supertypes are not being shown I still need to watch them.

Not sure how important it is to live update the subtype hierarchy.  The Java type hierarchy view does it.  But I don't know now to do it efficiently.   I'll have to experiment with this later. . . . So I should still branch this from 2435.

**Comment by AaronGreenhouse** [comment]
I don't need to search the whole workspace.  I only need to search the projects that recursively depend on the project in which the root classifier is located.  I'm not sure if this really helps or not.

**Comment by AaronGreenhouse** [comment]
It's probably a bad idea that the `AncestorTree` nodes hold on to `Classifier` nodes.  Also `MemberNode` elements hold on to `Element` and `Classifier` nodes.

**Comment by AaronGreenhouse** [comment]
As usual I nearly forgot about feature groups.  We also have

* feature group extension: i.e., `feature group A extends B` — feature group A extends feature group B — this is a `GroupExtension` node

* feature group inverse: i.e., `inverse of B` appears in feature group `A` — feature group A is an inverse of feature group B — this is a direct reference in the `FeatureGroupType` for A.

**Comment by AaronGreenhouse** [comment]
N.B. Need to refresh the descendants tree whenever a file in the project dependency tree changes.

**Comment by AaronGreenhouse** [comment]
Fixed the AncestorTree and the MemberTree to not hold onto EObject nodes.

**Comment by AaronGreenhouse** [comment]
Still need to update the resource change handling to work correctly with the descendant tree.

I current have 3 panes: ancestor, descendant, member.  Lutz said he would prefer to switch between the ancestor and descendant trees, so I ahve to fix that.

**Comment by AaronGreenhouse** [comment]
The three-pane classifier information view: Gets kinda cramped on small screens.

![Screen Shot 2020-10-12 at 10.46.28 AM.png](https://images.zenhubusercontent.com/5cc7040c07a4ab52d27abe58/369c17f0-61ea-4ece-a04e-c1ef5e8d8272)

![Screen Shot 2020-10-12 at 10.46.36 AM.png](https://images.zenhubusercontent.com/5cc7040c07a4ab52d27abe58/a13fb8ac-581f-4c5c-854e-cd2839d0b8c3)

**Comment by lwrage** [comment]
Is there a way to allow the user to switch between a two and three column layout?

**Comment by AaronGreenhouse** [comment]
@lwrage Yes, I don't see why we couldn't toggle between two and three columns, and then in the two column view have a toggle to switch between ancestors and descendants.

**Comment by AaronGreenhouse** [comment]
Fixed the resource change listener to monitor for the descendants as well by triggering a refresh whenever a file in the project tree changes.

**Comment by AaronGreenhouse** [comment]
Fixed the refresh to happen in a separate job launched in such a way that the view is marked as busy then later as changed.  (Still need to fix the progress bar.)

Need to get the view's "progress service":

```
	final IWorkbenchSiteProgressService service = getSite().getService(IWorkbenchSiteProgressService.class);
```

Schedule using the service:

```
	service.schedule(refreshJob, 0, true);
```

The `true` is supposed to make the curser "half-busy", but it doesn't seem to do anything on my Mac.

Can mark as changed when done:

```
	service.warnOfContentChange();
```

**Comment by AaronGreenhouse** [comment]
Added "Clear View" button to the view's toolbar so that large models can be removed from the view if they become annoying.

**Comment by AaronGreenhouse** [comment]
Added Progress monitor to the AADLFinder methods so that I can better handle progress in the Classifier Info view.  Still need to handle "cancel" of the job.

**Comment by AaronGreenhouse** [comment]
Need to redo the "image" caching in the view because i'm probably leaking images.

Come back to this.

**Comment by AaronGreenhouse** [comment]
Need to add an option to suspend automatic refresh, and enable manual refresh to handle the case of big models.

**Comment by AaronGreenhouse** [comment]
Need a status line in the view that shows which classifier is being displayed, and whether a refressh is needed.  This is necessary because cancelling a refresh can cause the display to be cleared.

**Comment by AaronGreenhouse** [comment]
Added menu option to the view to disable automatic refresh.  When disabled, a refresh button on the toolbar will be enabled _when a refresh is required_.

**Comment by AaronGreenhouse** [comment]
Switched back to a 2-paned view, with ancestor and descendant toggle buttons on the toolbar.  It is too hard to have an option to switch between a 2- and 3-paned view.

**Comment by AaronGreenhouse** [comment]
Discussed it with @lwrage and we decided the view does not need a context menu that allows refocusing the view on a the selected classifier.  Adding this would lead down the rabbit hole of having to remember past classifiers on which the view was focused, and we don't want to do that.

So last item remaining is to update the help text.

**Comment by AaronGreenhouse** [comment]
Overlooked a lot in the resource listener needed to refresh.  

* First, every time the view jumps to the source text, the view refreshes.  This is caused by the adding/removing of markers to the source text file to perform the jump.  Obviously we can ignore *all* marker changes.

* We can ignore all changes to any resource that isn't an `IFile`.

* We need to handle the case of the resource that contains the classifier being viewed being deleted.  In this case the view needs to clear its input.

**Comment by AaronGreenhouse** [comment]
Just checking for the `IResourceChange.REMOVED` kind on the resource delta is good enough to catch

* The file being moved

* The file being deleted

* The file being renamed

* The containing project being closed

**Comment by AaronGreenhouse** [comment]
So I made these changes, and found out something annoying/interesting.  When a refresh is reacting to the removal of an `.aadl` file, building the descendant tree throws an exception.  The resource finder complains that it cannot find the recently deleted file.  

Upon close inspection, what was happening was notification of the `.aadl` file being deleted, the refresh of the view would run (and fail), and then another resource change notification comes in that `.aadlbin` files have been updated.  **These files are needed to for the resource finder.**

More testing shows that any interesting change to an `.aadl` file (content change, creation, deletion, rename, closing the project) results in changes to the `.aadlbin` file corresponding to that file.  

So, I need to
1. Just watch for changes to the `.aadlbin` files in the projects of interest.  (Marker changes to the `.aadl` file do not affect the `.aadlbin` file.). 
2. Check that the resource of the currently viewed URI exists in `refresh()` and handle the case that it doesn't, instead of trying to use the events to determine that the resource was deleted.
